src/message_pool/msg_chain.rs

```diff
@@ -538,24 +538,26 @@ where
         chains.drop_invalid(&mut node_vec);
     }
 
-    // link next pointers
-    for i in 0..node_vec.len() - 1 {
-        let k1 = node_vec.get(i).unwrap();
-        let k2 = node_vec.get(i + 1);
-        let n1 = chains.get_mut(*k1).unwrap();
-        n1.next = k2.cloned();
-    }
-
-    // link prev pointers
-    for i in (0..node_vec.len() - 1).rev() {
-        let k1 = node_vec.get(i);
-        let k2 = node_vec.get(i + 1).unwrap();
-        let n2 = chains.get_mut(*k2).unwrap();
-        n2.prev = k1.cloned();
+    if node_vec.len() > 1 {
+        for (&k1, &k2) in node_vec.iter().zip(node_vec.iter().skip(1)) {
+            // link next pointers
+            let n1 = chains
+                .get_mut(k1)
+                .ok_or_else(|| Error::Other(format!("{k1:?} should present in `chains`")))?;
+            n1.next = Some(k2);
+            // Should we link or clear n1.prev as well?
```

Comment:
@aatifsyed Should we link or clear `n1.prev` and `n2.next` as well?

Reply:
I don't really understand this code at a first pass :/

Could you help me understand what these pointers mean from a domain perspective, or point me to some docs/pretty pictures that explain?

Reply:
@aatifsyed Sorry I might misread the git history and thought you were the author of this code. The logic 100% matches logic as I just checked so I guess it's fine although I have no idea how this linked list work either : )